impl/src/expand.rs

```diff
@@ -114,6 +118,48 @@ fn impl_struct(input: Struct) -> TokenStream {
         None
     };
     let display_impl = display_body.map(|body| {
+        let mut extra_predicates = Vec::new();
+        for field in input
+            .attrs
+            .display
+            .iter()
+            .flat_map(|d| d.iter_fmt_types(input.fields.as_slice()))
+        {
```

Comment:
In general these changes in expand.rs are way more code than I would expect to need to add for this feature. Please see #148 for the approach I used instead.

Reply:
Your `Field.contains_generic` approach combined with the `scope` abstraction seem to be the biggest contributors to code reduction, and `InferredBounds` is doing some serious work considering its length.

I'm curious, however, if there's some way to describe the "intent" for each bound as it's read, allowing a single `InferredBounds`-esque type to track the various usage forms to generalize the solution beyond `error` and `display` bounds.

Reply:
Yes that is how https://github.com/dtolnay/reflect works.

Reply:
I'll keep an eye out for opportunities to use that library- but I suspect it would be a bit much in the way of overhead for this task? Then again, being a proc macro, its dependencies (hopefully?) aren't compiled into the binary it helps to produce unless explicitly requested.